mount/upgrader_test.go

```diff
@@ -1,9 +1,270 @@
+			initial: carv1path,
+
+			createMnt: func(t *testing.T, key shard.Key, rootDir string) Mount {
+				return &FSMount{testdata, carv2path} // purposely giving a different file here.
+			},
+
+			verify: func(t *testing.T, u *Upgrader, key shard.Key, rootDir string) {
+				_, err := os.Stat(u.transientFilePath())
+				require.NoError(t, err)
+
+				// read the contents of the transient file.
+				tf, err := os.Open(u.transientFilePath())
+				require.NoError(t, err)
+				defer tf.Close()
+				bz, err := ioutil.ReadAll(tf)
+				require.NoError(t, err)
+				require.NoError(t, tf.Close())
+
+				// read the contents of the initial file -> they should match.
+				f, err := os.Open(carv1path)
+				require.NoError(t, err)
+				defer f.Close()
+				bz2, err := ioutil.ReadAll(f)
+				require.NoError(t, err)
+				require.NoError(t, f.Close())
+				require.EqualValues(t, bz, bz2)
+			},
+			expectedContentFilePath: carv1path,
+		},
+	}
+
+	ctx := context.Background()
+
+	for name, tc := range tcs {
+		tcc := tc
+		t.Run(name, func(t *testing.T) {
+			key := shard.KeyFromString(fmt.Sprintf("%d", rand.Uint64()))
+			rootDir := t.TempDir()
+			if tcc.setup != nil {
+				tcc.setup(t, key, rootDir)
+			}
+
+			mnt := tcc.createMnt(t, key, rootDir)
+
+			u, err := Upgrade(mnt, rootDir, key, tcc.initial)
+			require.NoError(t, err)
+			require.NotNil(t, u)
+			tcc.verify(t, u, key, rootDir)
+
+			// fetch and verify contents
+			rd, err := u.Fetch(ctx)
+			require.NoError(t, err)
+			require.NotNil(t, rd)
+			bz, err := ioutil.ReadAll(rd)
+			require.NoError(t, err)
+			require.NotEmpty(t, bz)
+			require.NoError(t, rd.Close())
+
+			f, err := os.Open(tcc.expectedContentFilePath)
+			require.NoError(t, err)
+			bz2, err := ioutil.ReadAll(f)
+			require.NoError(t, err)
+			require.NoError(t, f.Close())
+			require.EqualValues(t, bz2, bz)
+
+			// no transient file should exist as it's cleaned up when reader is closed and ref count drops to zero.
+			_, err = os.Stat(u.transientFilePath())
+			require.Error(t, err)
+
+		})
+	}
+}
+
+type countingMount struct {
+	mu sync.Mutex
+	n  int
+
+	Mount
+}
+
+func (c *countingMount) Fetch(ctx context.Context) (Reader, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.n++
+
+	return c.Mount.Fetch(ctx)
+}
+
+func (c *countingMount) Count() int {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.n
+}
 
-// TestUpgraderDeduplicatesRemote tests that when a remote has been fetched
-// successfully, and the local file is present, the remote is not hit again.
 func TestUpgraderDeduplicatesRemote(t *testing.T) {
-	t.Skip("TODO")
+	ctx := context.Background()
+	mnt := &countingMount{Mount: &FSMount{testdata, carv2path}}
+
+	key := shard.KeyFromString(fmt.Sprintf("%d", rand.Uint64()))
+	rootDir := t.TempDir()
+	u, err := Upgrade(mnt, rootDir, key, "")
+	require.NoError(t, err)
+	require.Zero(t, mnt.Count())
+
+	// now fetch in parallel
+	cnt := 20
+	var mu sync.Mutex
+	readers := make([]Reader, 0, cnt)
```

Comment:
No need to protect with a mutex if you size the slice to len 20 and you update the corresponding positions in the goroutine.